source-academy/cadet-frontend
# Fix monospace font inconsistencies

## Fix monospace font inconsistencies
Summary: Minor UX change.

### Description
Currently, [SICP JS](https://sicp.comp.nus.edu.sg) uses the font Inconsolata as the monospace font for rendering Source programs in the web edition of the textbook. In https://github.com/source-academy/cadet-frontend/pull/767#issuecomment-515885192 and #926, the default monospace font for the editor, REPL and a few other components were both updated to Inconsolata, to unify with the look of SICP JS.

Since Inconsolata is not a system font, but is hosted open-source on Google Fonts, the browser first fetches the CSS stylesheet containing the `@font-face` declaration from Google Fonts at [this URL](https://fonts.googleapis.com/css?family=Inconsolata&display=optional), before subsequently downloading the actual font file of extension `.woff2` from Google Fonts later. This is all and good.

However, as the `font-display` property is set as `optional`, this only gives the custom font a 100ms [block period](https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-display) (where invisible text is rendered in place of the loading font). If the custom font loads within this first 100ms, it will be used by the browser unconditionally. Otherwise, the decision to use the custom font is left to the browser - it can decide either to (1) render a fallback font first and then swap it out when the custom font loads, or (2) to entirely not use the custom font at all in favour of the fallback font.

<img src="https://user-images.githubusercontent.com/44989315/90721176-d0b57900-e2ea-11ea-9a2e-8d3414c80fdf.png" height="200" >

Unfortunately, in the case of the Academy, the Inconsolata web font is only requested earliest about 1.7 seconds in after the page is accessed (see waterfall diagram below). As a result, whilst the Inconsolata font downloads and loads quickly after the request is made, both supported browsers (Firefox and Chrome) decide not to use it, and continue to render the page with the fallback monospace font instead.

![image](https://user-images.githubusercontent.com/44989315/90727819-76bab080-e2f6-11ea-9171-882cc9cb28af.png)

This is unintended - the Inconsolata font should take precedence over the fallback font on the initial load, unless it fails to be retrieved even after the page fully renders (about 2.2s in). Furthermore, since the Inconsolata font is cached, if a user then reloads the page with F5 (which does not invalidate the cache unlike Ctrl+F5), the browser subsequently decides to use the Inconsolata font instead of the fallback. This creates the perception that the font in the Academy is inconsistent when the user performs a normal refresh for any reason.

I propose modifying the value of `font-display` to `block` instead, to allow the page up to 3 seconds to load the Inconsolata font before the browser renders text in the fallback font instead. This will largely ensure the Inconsolata font is used uniformly throughout the Academy, regardless if the user refreshes the page. On much slower networks (e.g. mobile data), the browser may still decide between swapping the fonts out after the initial 3 seconds, or not using Inconsolata at all.

### Changelog
- Modify the value of `font-display` for the Inconsolata `@font-face` loaded from Google Fonts from `optional` to `block`

### Verification instructions
Please test on **Firefox** (v70 or later) and **Chrome** (v75 or later):

1. Navigate to any page that makes use of monospace font for code, e.g. the Playground, an assessment/submission
2. Perform a hard refresh on the page with Ctrl-F5 - the page should load with all code rendered in the Inconsolata font
3. Perform a normal refresh on that same page with F5 - you should not observe any changes in the font
   - The current behaviour (in the absence of this PR) involves a change in font from the browser-defined fallback monospace font to Inconsolata

Last updated 20 Aug 2020, 6:45PM
